Update readme file (#475)
trustlines-protocol/blockchain: README.md

    - [Blockchain Node](#blockchain-node)
    - [Netstats Client](#netstats-client)
    - [Monitor](#monitor)
    - [Bridge](#bridge)
  - [Setup Without Docker](#setup-without-docker)
- [Development](#development)
  - [Build Own Image](#build-own-image)
  - [Upload Image](#upload-image)
  - [Running Tests on Contracts](#running-tests-on-contracts)

## The Trustlines Blockchain Infrastructure

Nodes of the Trustlines Blockchain run various applications:

- The node of the blockchain itself
- The monitor that checks if validators act honestly
- The bridge between Ethereum and the Trustlines Blockchain (only run by validators)
- The netstats client to report the node state to `https://netstats.trustlines.foundation`(optional)

There are multiple ways to set each of these up. The most straightforward one by far is via our interactive quickstart
script. Finer control can be achieved by starting the components individually as docker containers. Finally, it is also
possible to avoid docker altogether and run everything directly on the host machine.

Before starting the installation process, please have a look at the system requirements and the note on backups.

### System Requirements

> Based on the experiences we have had on our long-running testnet, we recommend at least 2GB of memory and 10GB of SSD

Author: True, I guess it's something we should double check, but for now it's our best estimate. I suspect we're not too far off though.

> Based on the experiences we have had on our long-running testnet, we recommend at least 2GB of memory and 10GB of SSD

Reviewer: Okay. If I remember it correctly @schmir-at-bb was working on issue about resources anyways. Maybe he can then adjust the recommendation as well.
Currently we already allow the mainnet node, bridge client and monitor to use together up to 2GB. The tlbc node and netstats client (I don't count watchtower) have no limits so far.

> Based on the experiences we have had on our long-running testnet, we recommend at least 2GB of memory and 10GB of SSD

Reviewer 2: the live test servers had 4GB of main memory available. I would rather recommend 4GB of main memory to be on the safe side. If we like to recommend the 2GB of memory, we should at least test it.
 
the 10GB of SSD storage are too small. I've got a testnet node running that's already using 9.2GB of storage.

> Based on the experiences we have had on our long-running testnet, we recommend at least 2GB of memory and 10GB of SSD

Reviewer: Yes. I think the old calculations of Michael are not much worth anymore.
Thanks for your input Ralf!

storage.

Validators should make sure their node has a high uptime: Otherwise, they miss out on potential revenue and harm the
network by increasing average block times.

For block validation and creation, it is essential to make sure your host system has the correct time configuration. On
most systems, this should be the case by default. You can check the settings with `timedatectl` (look for
`"System clock synchronized: yes"`).

> For the quickstart and docker installation modes, [docker](https://docker.com) needs to be installed and configured.

Author: I'd rather not link to this directly as it contains a changing version number. But users should be capable to find it on their own, so I think it's fine.

> For the quickstart and docker installation modes, [docker](https://docker.com) needs to be installed and configured.

Reviewer: That is true.

Please refer to the official documentation and make sure your user is added to the `docker` user group if you cannot
access root permissions to run containers.

### Backups

> For validators it is crucial to safely back up their private key.

Author: I moved it up explicitly to make it more visible and because the reading paths diverge depending on the installation method, so few people will read until the end.

Since validators participated in the auction with the same key already, they should have backups at this point and most people won't have to use `scp`.

> For validators it is crucial to safely back up their private key.

Reviewer: Okay, I'm fine with that opinion. The key is the only thing that actually matters.

To fetch the most recent version of the quickstart script and run it, execute the following command on your machine:

```sh
$ wget -O quickstart.sh https://github.com/trustlines-protocol/blockchain/raw/master/quickstart.sh && bash quickstart.sh
```

The script is interactive and will ask you which components to set up. Once the setup is complete, the various
components will run in the background in the form of docker containers. Configuration and chain data can be found in
the `trustlines` directory placed in the current working directory.

Executing the script again is safe: No configuration will be overridden. This allows you to add components not
configured in earlier runs and will restart all containers.

### Setup With Docker

A more explicit way of setting up the various components is starting the docker containers manually as described here.
To keep commands as concise as possible, only the most basic options are provided. You might want to set additional
ones, e.g., container names or restart policies.

#### Blockchain Node

First of all, to conveniently allow other containers to connect to the node, create a docker network with

```sh

> $ docker network create network-laika

Author: Originally I just used the host network for everything to use as few commands as possible. Then I tried to add a network to fix the netstats bug which didn't work. Now I'm not sure what's better anymore.

Author: I'd like to hold on with naming changes here until we've sorted them out globally. Same with testnet vs mainnet. We'll have to make another run through at this point anyways..

Reviewer: Okay, that sounds reasonable. :+1:

non-validator node:

```sh
$ mkdir -p trustlines/databases trustlines/config trustlines/enode trustlines/shared
$ docker run -d --name laika-node --network network-laika -v $(pwd)/trustlines/databases:/data -v $(pwd)/trustlines/config:/config/custom -v $(pwd)/trustlines/enode:/config/network -v $(pwd)/trustlines/shared:/shared/ -p 30300:30300 trustlines/tlbc-testnet
```

If you are a validator, this sequence of commands will supply Parity with your keystore file, password, and address so
that it can produce blocks:

```sh
$ mkdir -p trustlines/databases trustlines/config/keys/Trustlines trustlines/enode trustlines/shared
$ cp /path/to/your/keystore/file.json trustlines/config/keys/Trustlines
$ echo "<passphrase_for_keystore_file>" > trustlines/config/pass.pwd
$ docker run -d --name laika-node --network network-laika -v $(pwd)/trustlines/databases:/data -v $(pwd)/trustlines/config:/config/custom -v $(pwd)/trustlines/enode:/config/network -v $(pwd)/trustlines/shared:/shared/ -p 30300:30300 -p 30300:30300/udp trustlines/tlbc-testnet --role validator --address MY_ADDRESS
```

#### Netstats Client

> The netstats client reports the state of your node to [https://laikanetstats.trustlines.foundation/]() that gives a

Author: see above note on test vs main net

rough overview over the current state of the network. It is a fully optional component which helps the community, at
the cost leaks of some leakage of private information.

To participate, you first need to request credentials managed by the Trustlines Foundation. Please email
`netstats@trustlines.foundation` to do so.

Next, create a file named `netstats-env` in the `trustlines` directory with the following contents:

```sh
WS_USER=username-as-provided-by-the-foundation
WS_PASSWORD=password-as-provided-by-the-foundation
INSTANCE_NAME=please-choose-a-nice-name-here
```

If you want to be publicly reported as a validator, add the following line at the end:

```sh
HIDE_VALIDATOR_STATUS=false
```

The netstats client has to connect to the blockchain node. Therefore, we have to find out the node's IP address with

```sh

> docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' laika-node

Author: this was before the bridge config fixes, so this wouldn't have worked. Now it should, yes

> docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' laika-node

Reviewer: So the reason is? Because this way is cumbersome and less obvious. In my opinion. Anyways it is still a working solution.

```

Now, the netstats client can be started with

```sh

> $ docker run --name netstats-client --network network-laika -e trustlines/netstats-env -e RPC_HOST=<node IP address> -e RPC_PORT=8545 trustlines/netstats-client:master

Author: This command doesn't work. The netstats client connects to the node just fine via HTTP, but then apparently tries to make a websocket request which fails. Which is weird, because as far as I do pretty much the same as the docker compose file.

Author: Now it seems to not even be able to make the HTTP request.

```

#### Monitor

The monitor watches the blockchain and makes sure that validators are online and do not equivocate. Every node in the
network should run it and users should regularly check for reports of misbehaving validators.

Assuming the blockchain node was configured as described above, this command will start the monitor:

```sh
$ mkdir -p trustlines/monitor/reports trustlines/monitor/state
$ docker run -d --name tlbc-monitor --network network-laika -v $(pwd)/trustlines/shared:/config -v $(pwd)/trustlines/monitor/state:/state -v $(pwd)/trustlines/monitor/reports:/reports trustlines/tlbc-monitor -c /config/trustlines-spec.json -r /reports -d /state -u http://<node IP address>:8545
```

#### Bridge

Validators of the Trustlines Blockchain have to run the bridge that converts TLN tokens on the Ethereum chain to TLC

> tokens on the Trustlines Blockchain. Non-validators should not run a bridge node.

Author: To me this is a purely technical distinction that shouldn't matter to the user. Ether I think is worse because it refers to the Ethereum chain.

The bridge requires an Ethereum mainnet node which can be a light client. To start one, execute

```sh
$ docker network create network-ethereum
$ mkdir -p trustlines/databases/goerli

> $ docker run -d --name goerli-node --network network-ethereum -v $(pwd)/trustlines/databases/goerli:/data/database parity/parity:stable --light --no-download --auto-update none --chain goerli --db-path /data/database --base-path /data --no-hardware-wallets --jsonrpc-apis safe --jsonrpc-hosts all --jsonrpc-cors all --jsonrpc-port 8545 --no-ipc --no-secretstore --no-color

Author: This fails with a permission denied error:

```
Error upgrading parity data: CannotWriteVersionFile(Os { code: 13, kind: PermissionDenied, message: "Permission denied" })
```

If I don't use a volume I get

```
Error upgrading parity data: CannotCreateConfigPath(Os { code: 13, kind: PermissionDenied, message: "Permission denied" })
```

Reviewer: Because of the shared volume. Have a look at [this comment](https://github.com/trustlines-protocol/blockchain/blob/0b72bd96d1f7987abfea2b1257743718453d7694/docker/docker-compose.yaml#L35)

Reviewer: Format as the review comments above...

Author: I've seen this comment, but I'm not sure if this is the issue. That's why I tried not using a volume at all and it still failed.

Also, if it is the issue, how can we fix it? We don't have a compose file here.

Reviewer: > That's why I tried not using a volume at all and it still failed.

So just as a standard node without a validator key?

> Also, if it is the issue, how can we fix it? We don't have a compose file here.
1. Compose files can't add options to a container that _Docker_ does not support. So you can just use `docker run --user...` herefore.
2. You can also just use (a) different volume(s) for this node. :man_shrugging:

Author: I tried (2), but it doesn't work. I get the same error message as above.

I've added the `--user root` option, but I don't think this is an acceptable solution. Root permissions should not be necessary to run a node.

```

Similar as for the Laika node, find its IP address using

```sh
docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' goerli-node

> ```

Reviewer: As before...

event_fetch_start_block_number = 4153205

[validator_private_key]
keystore_path = "/config/keys/Trustlines/<keystore_filename>.json
keystore_password_path = "/config/pass.pwd"
```

Note that they keystore path is not an actual path on the host machine, but rather in the bridge container we will create next. The container will have to connect to both of the docker networks and access the config directory. Therefore, the command looks like this:

```sh
$ docker run -d --name bridge-client --network network-laika --network network-ethereum -v $(pwd)/trustlines/config:/config -v $(pwd)/trustlines/bridge-config.toml:/config/bridge-config.toml trustlines/bridge-next:master -c /config/bridge-config.toml
```

### Setup Without Docker

We refer to the documentation of the individual components:

- [Parity](https://wiki.parity.io/Parity-Ethereum)
- [Netstats](https://github.com/trustlines-protocol/ethstats-client)
- [Monitor](https://github.com/trustlines-protocol/tlbc-monitor)
- [Bridge](https://github.com/trustlines-protocol/blockchain/tree/master/tools/bridge)

For the Laika node, make sure it uses the correct [chain spec file](), that the right TCP and UDP ports are used

> (30300), and that the JSON RPC APIs `web3`, `eth`, and `net` are open.

Reviewer: ```suggestion
(30300), and that the JSON RPC APIs `web3`, `eth`, and `net` are enabled.
```
To do not mix it with open the port. Just to make it foolproof regarding security.

### Setup With the Quickstart Script

The quickstart script will set up the blockchain node and the monitor as well as optionally the bridge and netstats
clients. It allows importing a private key in order to act as a validator. In addition, it will start a
[watchtower](https://hub.docker.com/r/containrrr/watchtower) to automatically update the Docker containers when newer versions
become available (e.g. for bug fixes or network forks).

To fetch the most recent version of the quickstart script and run it, execute the following command on your machine:

```sh
$ wget -O quickstart.sh https://github.com/trustlines-protocol/blockchain/raw/master/quickstart.sh && bash quickstart.sh
```

The script is interactive and will ask you which components to set up. Once the
setup is complete, the various components will run in the background in the form
of Docker containers. Configuration and chain data can be found in the
`trustlines` directory placed in the current working directory. It is possible
to customize the own setup by editing those configuration files. This goes for
the configuration of the different components, as well as the composition of the

> Docker containers. If an optional component has not been setup on an earlier

Reviewer 3: ```suggestion
Docker containers. If an optional component has not been set up on an earlier
```

run, it can be added later by executing the quickstart script again.

Executing the script again is safe: No configuration will be overridden without
asking, in case the user has changed them itself. If conflicting configuration
updates occur, they are shown to the user who can ask to see a diff of the
changes.

### Setup With Docker

A more explicit way of setting up the various components is starting the Docker containers manually as described here.
To keep commands as concise as possible, only the most basic options are provided. You might want to set additional

> ones, e.g., container names or restart policies. Notice that we intent to

Reviewer 3: ```suggestion
ones, e.g. container names or restart policies. Notice that we intent to
```

> rebuild to setup of the quickstart script within the following sections. It is

Reviewer 3: `Notice that we intent to rebuild to setup of the quickstart script within the following sections.`
I do not understand this sentence, if you are ok with removing it, please do.

Reviewer: Copy of our _RocketChat_ conversation:
Hmm. My problem was that the descriptions by Jannik where quite “complicted” without a hint why they are like that. So he reconstructs the `trustlines` directory structure and all that. So in fact he is doing what the quickstart script is doing, but the manual way.

meant to get an idea how the components work and interact with each other. For
a complete setup you could also first use the quickstart and adjust the setup
afterwards, as it is fully customizable.

#### Blockchain Node

> The blockchain image is a standard Parity client but with a custom configuration for the Trustlines Blockchain. It also

Reviewer 3: ```suggestion
The blockchain image is a standard Parity client with a custom configuration for the Trustlines Blockchain. It also
```

    -v $(pwd)/trustlines/enode:/config/network \
    -v $(pwd)/trustlines/shared:/shared/ \
    -p 30300:30300 -p 30300:30300/udp \
    trustlines/tlbc-testnet
```

If you are a validator, this sequence of commands will supply Parity with your keystore file, password, and address so
that it can produce blocks:

```sh
$ mkdir -p trustlines/data-laika trustlines/config/keys/Trustlines trustlines/enode trustlines/shared
$ cp /path/to/your/keystore/file.json trustlines/config/keys/Trustlines
$ echo "<passphrase_for_keystore_file>" > trustlines/config/pass.pwd
$ docker run -d --name laika-node --network network-laika \
    -v $(pwd)/trustlines/data-laika:/data \
    -v $(pwd)/trustlines/config:/config/custom \
    -v $(pwd)/trustlines/enode:/config/network \
    -v $(pwd)/trustlines/shared:/shared/ \
    -p 30300:30300 -p 30300:30300/udp \
    trustlines/tlbc-testnet
```

#### Netstats Client

The netstats client reports the state of your node to the
[Laika netstats page](https://laikanetstats.trustlines.foundation/) that gives a rough overview of the current network state.
It is a fully optional component which helps the community, at the cost leaks of some leakage of

> private information.

Reviewer 3: In correct English: `It is a fully optional component which helps the community at the cost of leakage of some private information.`
I would simplify it as: `It is an optional component which helps the community by providing information on your running node to a central server.`

The bridge requires an Ethereum mainnet node which can be a light client. To start one, execute

```sh
$ docker network create network-ethereum
$ mkdir -p trustlines/data-goerli
$ docker run -d --name goerli-node --network network-ethereum \
    -v $(pwd)/trustlines/data-goerli:/data/database \
    --user root \
    ethereum/client-go:stable \
    --rpc --rpcaddr 0.0.0.0 --nousb --ipcdisable --goerli \
    --syncmode light --datadir /data/database --rpccorsdomain * \
    --rpcvhosts=*
```

Now, write a configuration file for the bridge node and store it in `trustlines/bridge-config.toml`:

```
[foreign_chain]
rpc_url = "http://goerli-node:8545"
token_contract_address = "0x54B06531214AD41DE9d771c10C0030d048d0cC67"
bridge_contract_address = "0x2171Dd4d4F6ca30FeEA8a27b96257A67f371d87A"
event_fetch_start_block_number = 1321331

[home_chain]
rpc_url = "http://laika-node:8545"
bridge_contract_address = "0x854dF872BB8bfECafFB1077FCfd7aa0B7C838A60"
event_fetch_start_block_number = 4153205

[validator_private_key]
keystore_path = "/config/keys/Trustlines/<keystore_filename.json>"
keystore_password_path = "/config/pass.pwd"
```

> Note that they keystore path is not an actual path on the host machine, but rather in the bridge container we will

Reviewer 3: ```suggestion
Note that the keystore path is not an actual path on the host machine, but rather in the bridge container we will
```

> Note that they keystore path is not an actual path on the host machine, but rather in the bridge container we will

Reviewer 3: ```suggestion
Note that the keystore path is not an actual path on the host machine, but rather in the bridge container we will
```
